### ISSUE TITLE
Improve internal logging layout

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#2506]
Goal: better readable, less noise

- level in caps
- remove logging of config rules 
- separator between message, level etc?
- other changes, to be determined

**Comment by FrancisChung** [comment]
Hi, I've been playing with some experimental hacks for a few hours and I think the following format might be a good starting point? (Module.MethodName would be NLog's Module.MethodName where the Logging ocurred)

```
2024-03-05 17:18:27.4062 [DEBUG] [Module.MethodName] Message Template Auto Format enabled
2024-03-05 17:18:27.4581 [TRACE] [Module.MethodName] FindReachableObject<NLog.Internal.IRenderable>
```
I think this would give people a good starting point for any place internally they want to dig deeper.

Unfortunately, there's no easy way to derive the Module.MethodName using Reflection or StackTrace objects in .NET Standard as it's not really supported (unlike 4.6 for example)

Another way of doing it could be using Aspect Orientated Programming as outlined [here](https://github.com/FrancisChung/CSharp-AOP-Logging-Primer) , but looking at how clean NLog's dependencies are, I'm guessing you wouldn't want to introduce any 3rd party dependencies.

**Comment by snakefoot** [comment]
Think NLog InternalLogger should be as simple as possible, and with minimum overhead.

There is already now a "hack" in NLog v5 where one can pass `this` as parameter, and then InternalLogger will extract context from first-parameter.

**Comment by FrancisChung** [comment]
@snakefoot Thanks for the heads up on the This parameter, IInternalLoggerContext and InternalLoggerMessageEventArgs. 
I'll take a look at those as well as #5443.

**Comment by FrancisChung** [comment]
I have managed to get some of the logs to utilise the ILoggingContext where available but there are plenty of calls where I'm not sure how to grab it from context. 

```
2024-03-06 15:31:24.7442 [ TRACE]  Scanning LiteralLayoutRenderer 'Layout Renderer: ${literal}'
2024-03-06 15:31:24.7442 [ TRACE]  Scanning MessageLayoutRenderer 'Layout Renderer: ${message}'
2024-03-06 15:31:24.7595 [ TRACE] [console|NLog.Targets.ColoredConsoleTarget] Initializing ColoredConsoleTarget(Name=console)
2024-03-06 15:31:24.7595 [ TRACE] FindReachableObject<NLog.Layouts.Layout>:
```

Any pointers would be greatly appreciated.

**Comment by FrancisChung** [comment]
I'm also playing with a multi-line indenter that is currently outputting a long line into something shorter
```

2024-03-06 15:31:24.7595 [ TRACE] Scanning Property RowHighlightingRules
                                     'System.Collections.Generic.List`1[NLog.Targets.ConsoleRowHighlightingRule]'
                                     System.Collections.Generic.IList`1[NLog.Targets.ConsoleRowHighlightingRule]
```

**Comment by snakefoot** [comment]
I don't have any pointers regarding InternalLogger. I just modify/improve error-texts when I see cases where it could give more light in the darkness. Small incremental steps to improve the current state, not trying to predict future needs. Again NLog has been written by programmers, so their technical-writing skills are very limited, and maybe english was not their native language. Most issues I have found about NLog InternalLogger is either completely forgetting the output for interesting events, or writing very vague error-texts that fails to explain the issue.

Regarding any formatting then #5443 makes it the task of the subscriber to perform any formatting.  Again the NLog InternalLogger is only something standard NLog-users look at to find errors. So it has to be extremely light-weight for all debug/info-output, and provide good hints on warnings/errors (Ex. what-to-do-details). But no one will be mapping back to source-code, besides core NLog-users maybe searching for a verbatim error-text. 

I only ask for debug-level from users that have environment-issue to figure out what is going on their computer, but no one except core NLog-users will have an easy time understanding the debug-output, and whether some debug-output-events are unexpected/interesting.

The original-issue-post is very broad, and I don't like broad issues. As one suddenly have to decipher what another person has been thinking. And the acceptance critera is also becomes vague. I think you should pick one simple thing to fix, and not try to solve all things at once.

**Comment by FrancisChung** [comment]
@snakefoot, I have definitely noticed that the internal logger's logs are an incoherent collection of entries that is confusing and cognitively heavy to decipher.

I think I will take out the multi-line indenter out for now, and concentrate on the readabilty of the logs. The logs doesn't have to be a Nobel Prize winning literature, but it should at least read out like a story.

**Comment by FrancisChung** [comment]
Would love some second opinion here on the formatting. I was thinking of full NameSpace names for Debug/Trace entries and shorter names for Info entries and Up, as illustrated below

```
2024-03-07 14:46:44.3589 [TRACE ] [Nlog.Internal.ObjectGraphScanner] Scanning MessageLayoutRenderer 'Layout Renderer: ${message}'
2024-03-07 14:46:44.3589 [DEBUG ] [NLog.Config.LoggingConfiguration] Adding target NLog.Targets.ColoredConsoleTarget(Name=console)
2024-03-07 14:46:44.6478 [INFO  ] [LoggingConfiguration] Registered target NLog.Targets.ColoredConsoleTarget(Name=console)
2024-03-07 14:46:44.6478 [TRACE ] [NLog.Config.LoggingConfigurationParser] ParseRulesElement() Entered
```

**Comment by snakefoot** [comment]
Again trying to predict future usually ends up at bad place with lots of complexity, that no one needs. I spend most of my time on improving `throwConfigExceptions="true"`, as I can see it really helps people (and me) by giving one clear error-message, and not a flood of output to digg throughy.

I don't have strong oppinions about the formatting of InternalLogger-output. It is a last resort place to look for clues, so any output is good output. You know that you are in a bad place when you have to crawl through the NLog InternalLogger-output :)

Again I don't have any huge need for the output to include internal NLog-class-names, as I usually just search the code for verbatim log-message. But if you have a smart system that has no overhead, and doesn't require people to explicit add NLog-class-names to the actual output, then I guess you can introduce a new formatting-option similar to IncludeTimestamp.

**Comment by FrancisChung** [comment]
> "Again trying to predict future usually ends up at bad place with lots of complexity, that no one needs."

I agree with you on this on principle, however that wasn't what I've done. As a complete noob on NLog that really haven't got too much clue on what is going on, the way I evolved the internal log so far has been very helpful in identifying what is being called when/where and sometimes how.

> "You know that you are in a bad place when you have to crawl through the NLog InternalLogger-output"

Pity the fool that volunteered to jump straight in here :)

> "Again I don't have any huge need for the output to include internal NLog-class-names, as I usually just search the code for verbatim log-message."

I've been doing this with my current small sample of log entries with mixed results, which is one of the reasons why I came up with the NLog-Classname labeling. Some of the log entries were not straight-forward searches.

> But if you have a smart system that has no overhead, and doesn't require people to explicitly add NLog-class-names to the actual output

I was initially hopeful that passing "This" and using IInternalLoggerContext would be the smart system that I could utilise, but in reality very few log entries were able to decipher the above correctly and output the correct desired output of NLog-classnames. There was only 2 log entries in my test run for example.

I've been explicitly adding them for now until now because of the above.
I understand your concerns about programmers being lazy and won't be adding 
I'm happy to take this onboard personally for the entirety of  NLog (in tranches), and happy to refactor if indeed the above "smart system" becomes a working reality within NLog.

I also think that once you've set overhauled and set a new standard on how internal logs look, most developers will adhere to the new standard. And arguably, the lazy attitude regarding internal logs is what got us in this mess in the first place

**Comment by FrancisChung** [comment]
Info or above level logs would have something a bit cleaner, like this

```
2024-03-08 10:38:59.4975 [INFO  ] [LoggingConfiguration] Registered target NLog.Targets.ColoredConsoleTarget(Name=console)
2024-03-08 10:38:59.8343 [INFO  ] [LoggingConfiguration]Validatingconfig:TargetNames=console,ConfigItems=9,
                                     FilePath=F:\Source\NLog\src\NLogConsole\bin\Debug\net8.0\NLog.config
2024-03-08 10:38:59.8343 [WARN  ] [LoggingConfiguration]Unusedtargetdetected.Addaruleforthistargettotheconfiguration.TargetName:
                                     console
2024-03-08 10:38:59.8415 [INFO  ] [LogFactory] Configurations for all loggers reloaded.
2024-03-08 10:38:59.8415 [INFO  ] [LogManager] Retrieved Logger for #Program
2024-03-08 11:49:19.6421 [INFO  ] [LogFactory] AppDomain Shutting down. LogFactory closing...
2024-03-08 11:49:19.6714 [INFO  ] [LogFactory] LogFactory has been closed.
```

**Comment by snakefoot** [comment]
I kind of like that the NLog InternalLogger format matches the NLog Default Layout (using pipes `|` instead of square brackets)
```
${longdate}|${level:uppercase=true}|${message:withexception=true}
```
Still not sure about explicit including ClassNams / ModuleName / AreaName / PrefixName in the Logger-output (Think it is better without full namespace). For normal NLog users it will just be noise, but it might be useful for new NLog developers navigating the NLog-source-code.

> Some of the log entries were not straight-forward searches.

Usually my approach has been to adjust the logging-statement, so it is be easier to find the next time. But I recognize that having AreaName in the logging-statement makes the code-search simpler.

At the same time I like NLog InternalLogger to be as fast and simple as possible without too many bells and whistles. Notice that some logging has been split into multiple logging-statements to avoid params-array allocations.
